### ISSUE TITLE
[red-knot] False-positive errors in `finally` suites when all other branches of the `try` block have terminal statements

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ruff#17016]
### Summary

Red-knot currently doesn't recognise any symbols in `finally` suites as having visible definitions if the `try` suite has a terminal statement and there either is no `except` suite OR the `except` suite also has terminal statements. This causes quite a few false positives when running red-knot on Black (for example).

```py
def a():
    try:
        return
    except:
        return
    finally:
        f = "foo"
        print(f)  # error: Name `f` is not defined

def b():
    try:
        return
    except:
        return
    finally:
        f = "foo"
        print(f)  # error: Name `f` is not defined


def c():
    try:
        return
    else:
        pass
    finally:
        f = "foo"
        print(f)  # error: Name `f` is not defined


def d():
    try:
        return
    except:
        return
    else:
        pass
    finally:
        f = "foo"
        print(f)  # error: Name `f` is not defined
```

There are no false positives on any of these:

```py
def e():
    try:
        pass
    except:
        return
    finally:
        f = "foo"
        print(f)

def f():
    try:
        return
    except:
        pass
    finally:
        f = "foo"
        print(f)

def g():
    try:
        pass
    except:
        pass
    else:
        return
    finally:
        f = "foo"
        print(f)

def h():
    try:
        return
    except:
        pass
    else:
        return
    finally:
        f = "foo"
        print(f)
```

**Comment by carljm** [comment]
This is purely a combination of https://github.com/astral-sh/ruff/issues/15797 and https://github.com/astral-sh/ruff/issues/14015.

A `finally` suite can run either as a result of an exception, or as a result of fall-through from `try` or `except` blocks. We currently model only the latter possibility.

And in the cases shown here, the latter possibility is unreachable, so this triggers our currently-suboptimal handling of unreachable code.

**Comment by AlexWaygood** [comment]
Hmm, I see. I suppose this simply validates our decision to treat #15797 as high priority, then :-) I'll close this out in favour of those two issues.

FWIW: this appears to cause about 20 false-positive diagnostics when running red-knot on Black
